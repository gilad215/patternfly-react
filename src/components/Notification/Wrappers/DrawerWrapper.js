import React from 'react';
import PropTypes from 'prop-types';
import { NotifDrawer } from '../NotifDrawer/index';
import PanelWrapper from './PanelWrapper';

const DrawerWrapper = ({
  notificationPanels,
  toggleDrawerHide,
  toggleDrawerExpand,
  isExpanded,
  togglePanel,
  onNotificationClick,
  onMarkPanelAsRead,
  onMarkPanelAsClear,
  onClickedLink
}) => {
  const panels = notificationPanels.map((panel, i) => (
    <PanelWrapper
      key={i}
      panelName={panel.panelName}
      panelkey={panel.panelkey}
      onClickedLink={onClickedLink}
      notifications={panel.notifs}
      togglePanel={togglePanel}
      isExpanded={panel.isExpanded}
      onNotificationClick={onNotificationClick}
      onMarkPanelAsRead={onMarkPanelAsRead}
      onMarkPanelAsClear={onMarkPanelAsClear}
      showLoading={panel.showLoading}
    />
  ));
  const noNotificationsMessage = (
    <div id="no-notifications-container">No Notifications</div>
  );

  return (
    <NotifDrawer expanded={isExpanded}>
      <NotifDrawer.Title
        title="Notifications"
        onCloseClick={() => toggleDrawerHide()}
        onExpandClick={toggleDrawerExpand}
      />
      <NotifDrawer.Accordion>
        {panels.length === 0 ? noNotificationsMessage : panels}
      </NotifDrawer.Accordion>
    </NotifDrawer>
  );
};

DrawerWrapper.propTypes = {
  /** toggleDrawer Hide/Expand func */
  toggleDrawerExpand: PropTypes.func,
  toggleDrawerHide: PropTypes.func,
  /** Notification Panels Array */
  notificationPanels: PropTypes.array,
  /** is Expanded Bool */
  isExpanded: PropTypes.bool,
  /** on Notification Click Func */
  onNotificationClick: PropTypes.func,
  /** on MarkAllRead Click Func */
  onMarkPanelAsRead: PropTypes.func,
  /** on Dropdown Link Click Func */
  onClickedLink: PropTypes.func,
  /** on MarkAllClear Click Func */
  onMarkPanelAsClear: PropTypes.func,
  /** on togglePanel Click Func */
  togglePanel: PropTypes.func
};

DrawerWrapper.defaultProps = {
  notificationPanels: null,
  toggleDrawerHide: null,
  toggleDrawerExpand: null,
  togglePanel: null,
  isExpanded: false,
  onNotificationClick: null,
  onMarkPanelAsRead: null,
  onClickedLink: null,
  onMarkPanelAsClear: null
};

export default DrawerWrapper;
