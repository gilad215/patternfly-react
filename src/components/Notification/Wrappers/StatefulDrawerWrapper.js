import React from 'react';
import PropTypes from 'prop-types';
import DrawerWrapper from './DrawerWrapper';

class StatefulDrawerWrapper extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      panels: props.panels,
      isExpanded: props.isExpanded
    };
  }

  onMarkPanelAsRead = panelkey => {
    const panels = this.state.panels.map(panel => {
      if (panel.panelkey === panelkey) {
        panel.notifs.map(notification => {
          notification.unread = false;
          return notification;
        });
      }
      return panel;
    });
    this.setState({ panels });
    this.updateUnreadCount();
  };

  onMarkPanelAsClear = key => {
    const panels = this.state.panels.map(panel => {
      if (panel.panelkey === key) panel.notifs = [];
      return panel;
    });
    this.setState({ panels });
    this.updateUnreadCount();
  };

  onNotificationClick = (panelkey, nkey) => {
    const panels = this.state.panels.map(panel => {
      if (panel.panelkey === panelkey) {
        panel.notifs.map(notification => {
          if (notification.nkey === nkey) notification.unread = false;
          return notification;
        });
      }
      return panel;
    });
    this.setState({ panels });
    this.updateUnreadCount();
  };

  togglePanel = key => {
    const panels = this.state.panels.map(panel => {
      if (panel.panelkey === key) panel.isExpanded = !panel.isExpanded;
      else panel.isExpanded = false;
      return panel;
    });
    this.setState({ panels });
  };

  toggleDrawerExpand = () => {
    this.setState(prevState => ({
      isExpanded: !prevState.isExpanded
    }));
  };

  updateUnreadCount = () => {
    let hasunread = false;
    for (let i = 0; i < this.state.panels.length; i++) {
      for (let j = 0; j < this.state.panels[i].notifs.length; j++) {
        if (this.state.panels[i].notifs[j].unread === true) {
          hasunread = true;
        }
      }
    }
    this.props.updateUnreadCount(hasunread);
  };

  render() {
    return (
      <DrawerWrapper
        notificationPanels={this.state.panels}
        togglePanel={this.togglePanel}
        toggleDrawerExpand={this.toggleDrawerExpand}
        isExpanded={this.state.isExpanded}
        toggleDrawerHide={this.toggleDrawer}
        onNotificationClick={this.onNotificationClick}
        onMarkPanelAsClear={this.onMarkPanelAsClear}
        onMarkPanelAsRead={this.onMarkPanelAsRead}
      />
    );
  }
}

StatefulDrawerWrapper.propTypes = {
  /** Notification Panels Array */
  panels: PropTypes.array,
  /** is Expanded Bool */
  isExpanded: PropTypes.bool,
  /** updateUnreadCount Func */
  updateUnreadCount: PropTypes.func
};
StatefulDrawerWrapper.defaultProps = {
  panels: null,
  isExpanded: false,
  updateUnreadCount: null
};

export default StatefulDrawerWrapper;
