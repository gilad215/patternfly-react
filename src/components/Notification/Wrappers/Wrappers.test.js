import React from 'react';
import renderer from 'react-test-renderer';
import PanelWraper from './PanelWrapper';
import DrawerWrapper from './DrawerWrapper';
import StatefulDrawerWrapper from './StatefulDrawerWrapper';
import StatefulToggleDrawerWrapper from './StatefulToggleDrawerWrapper';
import { Icon } from '../../Icon';
import getIconClass from './Icon.consts';

const notification = [
  {
    nkey: '4',
    expanded: false,
    unread: true,
    icon: {
      name: 'ok',
      type: 'pf'
    },
    text: 'Success Notification',
    date: '3/31/16',
    time: '12:12:44 PM',
    dropdownid: '4',
    dropdown: [
      {
        id: '4',
        href: 'https://theforeman.org/blog',
        title: 'Dropdown Link'
      }
    ]
  }
];

const panel = [
  {
    panelkey: '1',
    panelName: 'Notification Tab 1',
    notifs: notification,
    isExpanded: true
  }
];

test('PanelWraper is working properly', () => {
  const component = renderer.create(
    <PanelWraper
      panelkey={panel.panelkey}
      notifications={notification}
      togglePanel={jest.fn()}
      onNotificationClick={jest.fn()}
      onMarkPanelAsRead={jest.fn()}
      onClickedLink={jest.fn()}
      onMarkPanelAsClear={jest.fn()}
      showLoading
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('DrawerWraper is working properly', () => {
  const component = renderer.create(
    <DrawerWrapper
      notificationPanels={panel}
      toggleDrawerHide={jest.fn()}
      toggleDrawerExpand={jest.fn()}
      togglePanel={jest.fn()}
      onNotificationClick={jest.fn()}
      onMarkPanelAsRead={jest.fn()}
      onMarkPanelAsClear={jest.fn()}
      onClickedLink={jest.fn()}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('StatefulDrawerWrapper is working properly', () => {
  const component = renderer.create(
    <StatefulDrawerWrapper panels={panel} isExpanded />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('StatefulToggleDrawerWrapper is working properly', () => {
  const component = renderer.create(
    <StatefulToggleDrawerWrapper
      panels={panel}
      isDrawerOpen
      hasUnreadMessages
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('getIconClass is working properly', () => {
  const component = renderer.create(
    <Icon type="pf" name={getIconClass('ok')} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
