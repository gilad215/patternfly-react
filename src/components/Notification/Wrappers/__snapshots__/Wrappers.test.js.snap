// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerWraper is working properly 1`] = `
<div
  className="drawer-pf drawer-pf-notifications-non-clickable"
>
  <div
    className="drawer-pf-title"
  >
    <a
      className="drawer-pf-toggle-expand fa fa-angle-double-left hidden-xs"
      onClick={[MockFunction]}
    />
    <a
      className="drawer-pf-close pficon pficon-close"
      onClick={[Function]}
    />
    <h3
      className="text-center"
    >
      Notifications
    </h3>
  </div>
  <div
    className="panel-group"
    id="notification-drawer-accordion"
  >
    <div
      className="panel panel-default expanded"
    >
      <div
        className="panel-heading"
        onClick={[Function]}
      >
        <h4
          className="panel-title"
        >
          <a
            className=""
          >
            Notification Tab 1
          </a>
        </h4>
        <span
          className="panel-counter"
        >
          1 New Event
        </span>
      </div>
      <div
        aria-expanded={null}
        className="panel-collapse collapse in collapse in"
        id="1"
      >
        <div
          className="panel-body"
        >
          <div
            className="drawer-pf-notification unread"
          >
            <div
              className="dropdown pull-right dropdown-kebab-pf dropdown btn-group"
            >
              <button
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle btn btn-link"
                disabled={false}
                id="4"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                type="button"
              >
                <span
                  aria-hidden={true}
                  className="fa fa-ellipsis-v"
                />
              </button>
              <ul
                aria-labelledby="4"
                className="dropdown-menu-right dropdown-menu"
                role="menu"
              >
                <li
                  className=""
                  role="presentation"
                  style={undefined}
                >
                  <a
                    href="#"
                    id="notification-kebab-0"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    Dropdown Link
                  </a>
                </li>
              </ul>
            </div>
            <span
              aria-hidden="true"
              className="pficon pficon-ok pull-left"
            />
            <div
              className="drawer-pf-notification-content"
            >
              <span
                className="drawer-pf-notification-message"
                onClick={[Function]}
              >
                Success Notification
              </span>
              <div
                className="drawer-pf-notification-info"
              >
                <span
                  className="date"
                >
                  3/31/16
                </span>
                <span
                  className="time"
                >
                  12:12:44 PM
                </span>
              </div>
            </div>
          </div>
          <div
            className="drawer-pf-action"
          >
            <div
              className="drawer-pf-action-link drawer-pf-action-link"
              data-toggle="mark-all-read"
              onClick={[Function]}
            >
              <button
                className="btn btn-link"
                disabled={false}
                type="button"
              >
                Mark All Read
              </button>
            </div>
            <div
              className="drawer-pf-action-link"
              data-toggle="clear-all"
              onClick={[Function]}
            >
              <button
                className="btn btn-link"
                disabled={false}
                type="button"
              >
                <span
                  aria-hidden="true"
                  className="pficon pficon-close"
                />
                Clear All
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PanelWraper is working properly 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
    onClick={[Function]}
  >
    <h4
      className="panel-title"
    >
      <a
        className="collapsed"
      />
    </h4>
    <span
      className="panel-counter"
    >
      1 New Event
    </span>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse"
    id="1"
  >
    <div
      className="panel-body"
    >
      <div
        className="drawer-pf-notification unread"
      >
        <div
          className="dropdown pull-right dropdown-kebab-pf dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle btn btn-link"
            disabled={false}
            id="4"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
            <span
              aria-hidden={true}
              className="fa fa-ellipsis-v"
            />
          </button>
          <ul
            aria-labelledby="4"
            className="dropdown-menu-right dropdown-menu"
            role="menu"
          >
            <li
              className=""
              role="presentation"
              style={undefined}
            >
              <a
                href="#"
                id="notification-kebab-0"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                Dropdown Link
              </a>
            </li>
          </ul>
        </div>
        <span
          aria-hidden="true"
          className="pficon pficon-ok pull-left"
        />
        <div
          className="drawer-pf-notification-content"
        >
          <span
            className="drawer-pf-notification-message"
            onClick={[Function]}
          >
            Success Notification
          </span>
          <div
            className="drawer-pf-notification-info"
          >
            <span
              className="date"
            >
              3/31/16
            </span>
            <span
              className="time"
            >
              12:12:44 PM
            </span>
          </div>
        </div>
      </div>
      <div
        className="drawer-pf-notification text-center"
      >
        <div
          className="spinner spinner-xs spinner-inline"
        />
         Loading More
      </div>
      <div
        className="drawer-pf-action"
      >
        <div
          className="drawer-pf-action-link drawer-pf-action-link"
          data-toggle="mark-all-read"
          onClick={[Function]}
        >
          <button
            className="btn btn-link"
            disabled={false}
            type="button"
          >
            Mark All Read
          </button>
        </div>
        <div
          className="drawer-pf-action-link"
          data-toggle="clear-all"
          onClick={[Function]}
        >
          <button
            className="btn btn-link"
            disabled={false}
            type="button"
          >
            <span
              aria-hidden="true"
              className="pficon pficon-close"
            />
            Clear All
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StatefulDrawerWrapper is working properly 1`] = `
<div
  className="drawer-pf drawer-pf-notifications-non-clickable drawer-pf-expanded"
>
  <div
    className="drawer-pf-title"
  >
    <a
      className="drawer-pf-toggle-expand fa fa-angle-double-left hidden-xs"
      onClick={[Function]}
    />
    <a
      className="drawer-pf-close pficon pficon-close"
      onClick={[Function]}
    />
    <h3
      className="text-center"
    >
      Notifications
    </h3>
  </div>
  <div
    className="panel-group"
    id="notification-drawer-accordion"
  >
    <div
      className="panel panel-default expanded"
    >
      <div
        className="panel-heading"
        onClick={[Function]}
      >
        <h4
          className="panel-title"
        >
          <a
            className=""
          >
            Notification Tab 1
          </a>
        </h4>
        <span
          className="panel-counter"
        >
          1 New Event
        </span>
      </div>
      <div
        aria-expanded={null}
        className="panel-collapse collapse in collapse in"
        id="1"
      >
        <div
          className="panel-body"
        >
          <div
            className="drawer-pf-notification unread"
          >
            <div
              className="dropdown pull-right dropdown-kebab-pf dropdown btn-group"
            >
              <button
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle btn btn-link"
                disabled={false}
                id="4"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                type="button"
              >
                <span
                  aria-hidden={true}
                  className="fa fa-ellipsis-v"
                />
              </button>
              <ul
                aria-labelledby="4"
                className="dropdown-menu-right dropdown-menu"
                role="menu"
              >
                <li
                  className=""
                  role="presentation"
                  style={undefined}
                >
                  <a
                    href="#"
                    id="notification-kebab-0"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    Dropdown Link
                  </a>
                </li>
              </ul>
            </div>
            <span
              aria-hidden="true"
              className="pficon pficon-ok pull-left"
            />
            <div
              className="drawer-pf-notification-content"
            >
              <span
                className="drawer-pf-notification-message"
                onClick={[Function]}
              >
                Success Notification
              </span>
              <div
                className="drawer-pf-notification-info"
              >
                <span
                  className="date"
                >
                  3/31/16
                </span>
                <span
                  className="time"
                >
                  12:12:44 PM
                </span>
              </div>
            </div>
          </div>
          <div
            className="drawer-pf-action"
          >
            <div
              className="drawer-pf-action-link drawer-pf-action-link"
              data-toggle="mark-all-read"
              onClick={[Function]}
            >
              <button
                className="btn btn-link"
                disabled={false}
                type="button"
              >
                Mark All Read
              </button>
            </div>
            <div
              className="drawer-pf-action-link"
              data-toggle="clear-all"
              onClick={[Function]}
            >
              <button
                className="btn btn-link"
                disabled={false}
                type="button"
              >
                <span
                  aria-hidden="true"
                  className="pficon pficon-close"
                />
                Clear All
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StatefulToggleDrawerWrapper is working properly 1`] = `
<ul
  className="nav navbar-nav navbar-right navbar-iconic"
>
  <button
    className="btn btn-link"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    <span
      aria-describedby="tooltip"
      aria-hidden={true}
      className="fa fa-bell fa-lg"
      style={
        Object {
          "color": "rgb(240, 240, 240)",
          "marginTop": "20px",
        }
      }
    />
  </button>
  <div
    className="drawer-pf drawer-pf-notifications-non-clickable"
  >
    <div
      className="drawer-pf-title"
    >
      <a
        className="drawer-pf-toggle-expand fa fa-angle-double-left hidden-xs"
        onClick={[Function]}
      />
      <a
        className="drawer-pf-close pficon pficon-close"
        onClick={[Function]}
      />
      <h3
        className="text-center"
      >
        Notifications
      </h3>
    </div>
    <div
      className="panel-group"
      id="notification-drawer-accordion"
    >
      <div
        className="panel panel-default expanded"
      >
        <div
          className="panel-heading"
          onClick={[Function]}
        >
          <h4
            className="panel-title"
          >
            <a
              className=""
            >
              Notification Tab 1
            </a>
          </h4>
          <span
            className="panel-counter"
          >
            1 New Event
          </span>
        </div>
        <div
          aria-expanded={null}
          className="panel-collapse collapse in collapse in"
          id="1"
        >
          <div
            className="panel-body"
          >
            <div
              className="drawer-pf-notification unread"
            >
              <div
                className="dropdown pull-right dropdown-kebab-pf dropdown btn-group"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="dropdown-toggle btn btn-link"
                  disabled={false}
                  id="4"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  type="button"
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-ellipsis-v"
                  />
                </button>
                <ul
                  aria-labelledby="4"
                  className="dropdown-menu-right dropdown-menu"
                  role="menu"
                >
                  <li
                    className=""
                    role="presentation"
                    style={undefined}
                  >
                    <a
                      href="#"
                      id="notification-kebab-0"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      Dropdown Link
                    </a>
                  </li>
                </ul>
              </div>
              <span
                aria-hidden="true"
                className="pficon pficon-ok pull-left"
              />
              <div
                className="drawer-pf-notification-content"
              >
                <span
                  className="drawer-pf-notification-message"
                  onClick={[Function]}
                >
                  Success Notification
                </span>
                <div
                  className="drawer-pf-notification-info"
                >
                  <span
                    className="date"
                  >
                    3/31/16
                  </span>
                  <span
                    className="time"
                  >
                    12:12:44 PM
                  </span>
                </div>
              </div>
            </div>
            <div
              className="drawer-pf-action"
            >
              <div
                className="drawer-pf-action-link drawer-pf-action-link"
                data-toggle="mark-all-read"
                onClick={[Function]}
              >
                <button
                  className="btn btn-link"
                  disabled={false}
                  type="button"
                >
                  Mark All Read
                </button>
              </div>
              <div
                className="drawer-pf-action-link"
                data-toggle="clear-all"
                onClick={[Function]}
              >
                <button
                  className="btn btn-link"
                  disabled={false}
                  type="button"
                >
                  <span
                    aria-hidden="true"
                    className="pficon pficon-close"
                  />
                  Clear All
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</ul>
`;
